File: prototype/klondike-table/game/reducer.js
/** PROTOTYPE — Klondike reducer. Pure; portable into the real app later. */

import { dealGame } from './deal.js';
import {
  canStackOnFoundation,
  canStackOnTableau,
  getPile,
  isWin,
  tableauRunIsLegal,
} from './rules.js';

function cloneState(state) {
  return {
    stock: state.stock.map((c) => ({ ...c })),
    waste: state.waste.map((c) => ({ ...c })),
    foundations: state.foundations.map((p) => p.map((c) => ({ ...c }))),
    tableau: state.tableau.map((p) => p.map((c) => ({ ...c }))),
    selection: state.selection
      ? { ...state.selection, from: { ...state.selection.from } }
      : null,
    won: state.won,
  };
}

function setPile(state, pile, cards) {
  switch (pile.area) {
    case 'stock':
      state.stock = cards;
      break;
    case 'waste':
      state.waste = cards;
      break;
    case 'foundation':
      state.foundations[pile.index] = cards;
      break;
    case 'tableau':
      state.tableau[pile.index] = cards;
      break;
  }
}

function samePile(a, b) {
  if (a.area !== b.area) return false;
  if (a.area === 'foundation' && b.area === 'foundation') return a.index === b.index;
  if (a.area === 'tableau' && b.area === 'tableau') return a.index === b.index;
  return true;
}

function takeSelection(state, sel) {
  const pile = getPile(state, sel.from);
  if (sel.from.area === 'stock') return null;
  if (sel.from.area === 'waste' || sel.from.area === 'foundation') {
    if (pile.length === 0) return null;
    return [pile[pile.length - 1]];
  }
  if (!tableauRunIsLegal(pile, sel.cardIndex)) return null;
  return pile.slice(sel.cardIndex);
}

function removeSelection(state, sel, count) {
  const pile = getPile(state, sel.from).slice();
  if (sel.from.area === 'waste' || sel.from.area === 'foundation') {
    pile.pop();
  } else if (sel.from.area === 'tableau') {
    pile.splice(sel.cardIndex, count);
    if (pile.length > 0 && !pile[pile.length - 1].faceUp) {
      pile[pile.length - 1] = { ...pile[pile.length - 1], faceUp: true };
    }
  }
  setPile(state, sel.from, pile);
}

function canDrop(moving, onto, state) {
  if (moving.length === 0) return false;
  const head = moving[0];
  if (onto.area === 'stock' || onto.area === 'waste') return false;
  if (onto.area === 'foundation') {
    if (moving.length !== 1) return false;
    return canStackOnFoundation(head, state.foundations[onto.index]);
  }
  const targetPile = state.tableau[onto.index];
  const target = targetPile.length ? targetPile[targetPile.length - 1] : undefined;
  return canStackOnTableau(head, target);
}

function applyDrop(state, onto, from, cardIndex) {
  const next = cloneState(state);
  // Drag can name the source directly; tap→tap uses existing selection.
  if (from) {
    next.selection = { from, cardIndex: cardIndex ?? 0 };
  }
  if (!next.selection) return state;
  const sel = next.selection;
  if (samePile(sel.from, onto)) {
    next.selection = null;
    return next;
  }
  const moving = takeSelection(next, sel);
  if (!moving || !canDrop(moving, onto, next)) {
    next.selection = null;
    return next;
  }
  removeSelection(next, sel, moving.length);
  const dest = getPile(next, onto).slice();
  dest.push(...moving.map((c) => ({ ...c, faceUp: true })));
  setPile(next, onto, dest);
  next.selection = null;
  next.won = isWin(next.foundations);
  return next;
}

function applySelect(state, pile, cardIndex) {
  if (pile.area === 'stock') return state;
  const cards = getPile(state, pile);
  if (cards.length === 0) return { ...state, selection: null };

  if (pile.area === 'waste' || pile.area === 'foundation') {
    return { ...state, selection: { from: pile, cardIndex: cards.length - 1 } };
  }

  const idx = cardIndex ?? cards.length - 1;
  if (idx < 0 || idx >= cards.length || !cards[idx].faceUp) return state;
  if (!tableauRunIsLegal(cards, idx)) return state;
  return { ...state, selection: { from: pile, cardIndex: idx } };
}

function applyAutoMove(state, from, cardIndex) {
  if (from.area === 'stock') return state;
  const cards = getPile(state, from);
  if (cards.length === 0) return { ...state, selection: null };

  let idx = cardIndex;
  if (from.area === 'waste' || from.area === 'foundation') {
    idx = cards.length - 1;
  } else {
    idx = cardIndex ?? cards.length - 1;
    if (idx < 0 || idx >= cards.length || !cards[idx].faceUp) {
      return { ...state, selection: null };
    }
    if (!tableauRunIsLegal(cards, idx)) return { ...state, selection: null };
  }

  const moving = takeSelection(state, { from, cardIndex: idx });
  if (!moving) return { ...state, selection: null };

  // Prefer Foundations (e.g. 2 → matching Ace), then any legal Tableau.
  for (let i = 0; i < 4; i++) {
    const onto = { area: 'foundation', index: i };
    if (canDrop(moving, onto, state)) {
      return applyDrop(state, onto, from, idx);
    }
  }
  for (let i = 0; i < 7; i++) {
    const onto = { area: 'tableau', index: i };
    if (from.area === 'tableau' && from.index === i) continue;
    if (canDrop(moving, onto, state)) {
      return applyDrop(state, onto, from, idx);
    }
  }
  return { ...state, selection: null };
}

function applyTap(state, pile, cardIndex) {
  if (pile.area === 'stock') return reduce(state, { type: 'DRAW' });
  if (state.selection) return applyDrop(state, pile);
  return applySelect(state, pile, cardIndex);
}

function draw(state) {
  const next = cloneState(state);
  next.selection = null;
  if (next.stock.length === 0) {
    // Computer Klondike: recycle Waste → Stock (unlimited passes).
    // Flip the Waste pile face-down onto the Stock so drawing continues.
    if (next.waste.length === 0) return next;
    next.stock = next.waste
      .slice()
      .reverse()
      .map((c) => ({ ...c, faceUp: false }));
    next.waste = [];
    return next;
  }
  const card = next.stock.pop();
  next.waste.push({ ...card, faceUp: true });
  return next;
}

export function reduce(state, action) {
  switch (action.type) {
    case 'NEW_GAME':
      return dealGame(action.seed ?? Date.now());
    case 'DRAW':
      return draw(state);
    case 'TAP':
      return applyTap(state, action.pile, action.cardIndex);
    case 'SELECT':
      return applySelect(state, action.pile, action.cardIndex);
    case 'DROP':
      return applyDrop(state, action.onto, action.from, action.cardIndex);
    case 'AUTO_MOVE':
      return applyAutoMove(state, action.pile, action.cardIndex);
    case 'CLEAR_SELECTION':
      return { ...state, selection: null };
    default:
      return state;
  }
}

export function selectedCardIds(state) {
  const ids = new Set();
  if (!state.selection) return ids;
  const cards = takeSelection(state, state.selection);
  if (!cards) return ids;
  for (const c of cards) ids.add(c.id);
  return ids;
}

File: prototype/klondike-table-flutter/lib/game/reducer.dart
/// PROTOTYPE — Klondike reducer. Pure; portable into the real app later.
library;

import 'deal.dart';
import 'plays.dart';
import 'rules.dart';

sealed class GameAction {
  const GameAction();
}

class NewGameAction extends GameAction {
  const NewGameAction([this.seed, this.drawType = DrawType.drawOne]);
  final int? seed;
  final DrawType drawType;
}

class DrawAction extends GameAction {
  const DrawAction();
}

class TapAction extends GameAction {
  const TapAction(this.pile, [this.cardIndex]);
  final PileRef pile;
  final int? cardIndex;
}

class DropAction extends GameAction {
  const DropAction(this.onto, {this.from, this.cardIndex});
  final PileRef onto;
  final PileRef? from;
  final int? cardIndex;
}

class AutoMoveAction extends GameAction {
  const AutoMoveAction(this.pile, [this.cardIndex]);
  final PileRef pile;
  final int? cardIndex;
}

class ClearSelectionAction extends GameAction {
  const ClearSelectionAction();
}

class _Board {
  _Board(GameState state)
    : stock = [...state.stock],
      waste = [...state.waste],
      foundations = [
        for (final p in state.foundations) [...p],
      ],
      tableau = [
        for (final p in state.tableau) [...p],
      ],
      selection = state.selection,
      won = state.won,
      drawType = state.drawType,
      seenFaceUp = state.seenFaceUp;

  List<PlayingCard> stock;
  List<PlayingCard> waste;
  List<List<PlayingCard>> foundations;
  List<List<PlayingCard>> tableau;
  Selection? selection;
  bool won;
  DrawType drawType;
  Set<String> seenFaceUp;

  GameState freeze() => GameState(
    stock: stock,
    waste: waste,
    foundations: foundations,
    tableau: tableau,
    selection: selection,
    won: won,
    drawType: drawType,
    seenFaceUp: seenFaceUp,
  );

  List<PlayingCard> pileOf(PileRef pile) {
    switch (pile.area) {
      case PileArea.stock:
        return stock;
      case PileArea.waste:
        return waste;
      case PileArea.foundation:
        return foundations[pile.index];
      case PileArea.tableau:
        return tableau[pile.index];
    }
  }

  void setPile(PileRef pile, List<PlayingCard> cards) {
    switch (pile.area) {
      case PileArea.stock:
        stock = cards;
      case PileArea.waste:
        waste = cards;
      case PileArea.foundation:
        foundations[pile.index] = cards;
      case PileArea.tableau:
        tableau[pile.index] = cards;
    }
  }
}

List<PlayingCard>? _takeSelection(_Board state, Selection sel) {
  final pile = state.pileOf(sel.from);
  if (sel.from.area == PileArea.stock) return null;
  if (sel.from.area == PileArea.waste || sel.from.area == PileArea.foundation) {
    if (pile.isEmpty) return null;
    return [pile.last];
  }
  if (!tableauRunIsLegal(pile, sel.cardIndex)) return null;
  return pile.sublist(sel.cardIndex);
}

void _removeSelection(_Board state, Selection sel, int count) {
  final pile = [...state.pileOf(sel.from)];
  if (sel.from.area == PileArea.waste || sel.from.area == PileArea.foundation) {
    pile.removeLast();
  } else if (sel.from.area == PileArea.tableau) {
    pile.removeRange(sel.cardIndex, sel.cardIndex + count);
    if (pile.isNotEmpty && !pile.last.faceUp) {
      pile[pile.length - 1] = pile.last.copyWith(faceUp: true);
    }
  }
  state.setPile(sel.from, pile);
}

bool _canDrop(List<PlayingCard> moving, PileRef onto, _Board state) {
  if (moving.isEmpty) return false;
  final head = moving.first;
  if (onto.area == PileArea.stock || onto.area == PileArea.waste) return false;
  if (onto.area == PileArea.foundation) {
    if (moving.length != 1) return false;
    return canStackOnFoundation(head, state.foundations[onto.index]);
  }
  final targetPile = state.tableau[onto.index];
  final target = targetPile.isEmpty ? null : targetPile.last;
  return canStackOnTableau(head, target);
}

GameState applyDrop(
  GameState state,
  PileRef onto, [
  PileRef? from,
  int? cardIndex,
]) {
  final next = _Board(state);
  if (from != null) {
    next.selection = Selection(from: from, cardIndex: cardIndex ?? 0);
  }
  final sel = next.selection;
  if (sel == null) return state;
  if (sel.from.sameAs(onto)) {
    next.selection = null;
    return next.freeze();
  }
  final moving = _takeSelection(next, sel);
  if (moving == null || !_canDrop(moving, onto, next)) {
    next.selection = null;
    return next.freeze();
  }
  _removeSelection(next, sel, moving.length);
  final dest = [...next.pileOf(onto)];
  dest.addAll([for (final c in moving) c.copyWith(faceUp: true)]);
  next.setPile(onto, dest);
  next.selection = null;
  next.won = isWin(next.foundations);
  return next.freeze();
}

GameState applySelect(GameState state, PileRef pile, int? cardIndex) {
  if (pile.area == PileArea.stock) return state;
  final cards = getPile(state, pile);
  if (cards.isEmpty) {
    return state.copyWith(selection: null);
  }
  if (pile.area == PileArea.waste || pile.area == PileArea.foundation) {
    return state.copyWith(
      selection: Selection(from: pile, cardIndex: cards.length - 1),
    );
  }
  final idx = cardIndex ?? cards.length - 1;
  if (idx < 0 || idx >= cards.length || !cards[idx].faceUp) return state;
  if (!tableauRunIsLegal(cards, idx)) return state;
  return state.copyWith(
    selection: Selection(from: pile, cardIndex: idx),
  );
}

GameState applyAutoMove(GameState state, PileRef from, int? cardIndex) {
  if (from.area == PileArea.stock) return state;
  final cards = getPile(state, from);
  if (cards.isEmpty) {
    return state.copyWith(selection: null);
  }

  late final int idx;
  if (from.area == PileArea.waste || from.area == PileArea.foundation) {
    idx = cards.length - 1;
  } else {
    idx = cardIndex ?? cards.length - 1;
    if (idx < 0 || idx >= cards.length || !cards[idx].faceUp) {
      return state.copyWith(selection: null);
    }
    if (!tableauRunIsLegal(cards, idx)) {
      return state.copyWith(selection: null);
    }
  }

  final play = autoMovePlay(state, from, idx);
  if (play == null) {
    return state.copyWith(selection: null);
  }
  return applyDrop(state, play.onto, from, idx);
}

GameState applyTap(GameState state, PileRef pile, int? cardIndex) {
  if (pile.area == PileArea.stock) return reduce(state, const DrawAction());
  if (state.selection != null && state.selection!.from.sameAs(pile)) {
    final cards = getPile(state, pile);
    final idx = pile.area == PileArea.waste || pile.area == PileArea.foundation
        ? cards.length - 1
        : cardIndex ?? cards.length - 1;
    if (idx == state.selection!.cardIndex) {
      return state.copyWith(selection: null);
    }
    return applySelect(state, pile, cardIndex);
  }
  if (state.selection != null) return applyDrop(state, pile);
  return applySelect(state, pile, cardIndex);
}

GameState draw(GameState state) {
  final next = _Board(state);
  next.selection = null;
  if (next.stock.isEmpty) {
    if (next.waste.isEmpty) return next.freeze();
    next.stock = [
      for (final c in next.waste.reversed) c.copyWith(faceUp: false),
    ];
    next.waste = [];
    return next.freeze();
  }
  final n = state.drawType == DrawType.drawThree ? 3 : 1;
  final take = n < next.stock.length ? n : next.stock.length;
  final drawn = <PlayingCard>[];
  for (var i = 0; i < take; i++) {
    drawn.add(next.stock.removeLast().copyWith(faceUp: true));
  }
  next.waste = [...next.waste, ...drawn];
  return next.freeze();
}

GameState reduce(GameState state, GameAction action) {
  switch (action) {
    case NewGameAction(:final seed, :final drawType):
      return dealGame(
        seed: seed ?? DateTime.now().millisecondsSinceEpoch,
        drawType: drawType,
      );
    case DrawAction():
      return draw(state);
    case TapAction(:final pile, :final cardIndex):
      return applyTap(state, pile, cardIndex);
    case DropAction(:final onto, :final from, :final cardIndex):
      return applyDrop(state, onto, from, cardIndex);
    case AutoMoveAction(:final pile, :final cardIndex):
      return applyAutoMove(state, pile, cardIndex);
    case ClearSelectionAction():
      return state.copyWith(selection: null);
  }
}

Set<String> selectedCardIds(GameState state) {
  final ids = <String>{};
  final sel = state.selection;
  if (sel == null) return ids;
  final cards = _takeSelection(_Board(state), sel);
  if (cards == null) return ids;
  for (final c in cards) {
    ids.add(c.id);
  }
  return ids;
}
